src/runtime_src/core/common/memaccess.cpp

```diff
@@ -0,0 +1,216 @@
+
+  return start_bank;
+}
+
+static uint64_t
+get_available_memory_size(std::vector<mem_bank_t>& vec_banks, std::vector<mem_bank_t>::iterator& start_bank, const uint64_t start_addr) 
+{
+  // Validate the amount of accessable memory
+  uint64_t available_size = std::accumulate(start_bank, vec_banks.end(), (uint64_t)0,
+    [](uint64_t result, const mem_bank_t& obj) {
+      return result + obj.m_size;
+    }) ;
+
+  available_size -= start_addr - start_bank->m_base_address;
+
+  return available_size;
+}
+
+enum class operation_type {
+  read,
+  write
+};
+
+// Ensure safe access into a device's memory banks based on memory
+// bank boundary and if the bank is in use
+static void 
+perform_memory_action(xrt_core::device* device, xrt_core::aligned_ptr_type& buf, const uint64_t start_addr, const uint64_t size, operation_type action)
+{
+  auto vec_banks = get_ddr_banks(device);
+  auto validated_start_addr = get_starting_address(vec_banks, start_addr);
+  auto start_bank = get_starting_bank(vec_banks, validated_start_addr);
+  auto available_size = get_available_memory_size(vec_banks, start_bank, validated_start_addr);
+
+  // Validate the size of the memory operation
+  if (size > available_size)
+    throw xrt_core::error(std::errc::operation_canceled, boost::str(boost::format("Cannot access %d bytes of memory from start address 0x%x\n") % size % start_addr));
+
+  auto validated_size = size;
+  // If no size is specified for the read. Read all available memory
+  if ((size == 0) && (action == operation_type::read))
+    validated_size = available_size;
+
+  uint64_t current_addr = validated_start_addr;
+  size_t remaining_bytes_to_see = size;
+  size_t bytes_seen = 0;
+
+  // continue to operate as long as there are bytes left to see or we run out of banks
+  for (auto it = start_bank; (it != vec_banks.end()) && (remaining_bytes_to_see > 0); ++it) {
+    // Validate the amount of memory the current memory bank has
+    uint64_t available_bank_size = 0;
+    if (it != start_bank) {
+      current_addr = it->m_base_address;
+      available_bank_size = it->m_size;
+    }
+    else 
+      available_bank_size = it->m_size - (current_addr - it->m_base_address);
+
+    // If the available bank size is less than the source buffer see what bytes we are able to and move to the next bank
+    uint64_t bytes_to_edit = std::min(available_bank_size, validated_size);
+
+    // Update the buffer index based on how far we have read
+    void* current_buffer_location = static_cast<char *>(buf.get()) + bytes_seen;
+    switch (action) {
+      case operation_type::read:
+        device->unmgd_pread(current_buffer_location, bytes_to_edit, validated_start_addr);
+        break;
+      case operation_type::write:
+        device->unmgd_pwrite(current_buffer_location, bytes_to_edit, validated_start_addr);
+        break;
+      default:
```

Comment:
The compiler should detect that this is a switch.  If there are any missing enumeration values a warning / error should be produced.  As such, this default is not needed.